src/clients/FtpRelay.cc

```diff
@@ -385,9 +385,16 @@ Ftp::Relay::forwardReply()
     reply->sources |= Http::Message::srcFtp;
 
     setVirginReply(reply);
-    markParsedVirginReplyAsWhole("Ftp::Relay::handleControlReply() does not forward partial replies");
     adaptOrFinalizeReply();
 
+#if USE_ADAPTATION
+    if (adaptationAccessCheckPending) {
+        serverComplete();
+        debugs(9, 3, "returning due to adaptationAccessCheckPending");
+        return;
+    }
+#endif
+    markParsedVirginReplyAsWhole("custom/complete reply");
     serverComplete();
```

Comment:
Please ignore this change request if the bug ends up being fixed without modifying this method.

If PR changes to this method survive, can we simplify them by calling serverComplete() first, before dealing with adaptation and markParsedVirginReplyAsWhole()? After all, we are done working with the peer/server here, right? AFAICT, serverComplete2() will check doneWithAdaptation() and prevent this task from completing if we are still adapting.

```suggestion
    serverComplete();

#if USE_ADAPTATION
    if (!doneWithAdaptation()) {
        // XXX: Why should adaptation decide how to mark a _virgin_ response?!
        debugs(9, 3, "adaptation will decide whether to markParsedVirginReplyAsWhole()");
        return;
    }
#endif
    markParsedVirginReplyAsWhole("Ftp::Relay::handleControlReply() does not forward partial replies");
```

Reply:
> can we simplify them by calling serverComplete() first

We cannot call serverComplete() before markParsedVirginReplyAsWhole() because without adaptation FwdState::completed() (that will be called eventually) expects storedWholeReply_ set (otherwise it would call entry->completeTruncated()).